File: graph/easy/find_the_town_judge_997.go
package easy

func findJudge(n int, trust [][]int) int {
	citizenMap := make(map[int]map[int]interface{})

	citizen := 1

	for citizen <= n {
		citizenMap[citizen] = make(map[int]interface{})
		citizen++
	}

	for _, v := range trust {
		citizenMap[v[0]][v[1]] = nil
	}
	judge := -1 << 31

	for k, v := range citizenMap {
		if len(v) == 0 {
			judge = k
			break
		}
	}
	if judge == -1<<31 {
		return -1
	}
	for k, v := range citizenMap {
		if k == judge {
			continue
		} else {
			if _, ok := v[judge]; !ok {
				return -1
			}
		}
	}
	return judge
}

File: leetcode/array/easy/java_solutions/FindTheTownJudge_997.java
package leetcode.array.easy.java_solutions;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class FindTheTownJudge_997 {

    static class Solution {
        public int findJudge(int n, int[][] trust) {
            int[] counter = new int[n + 1];

            for (int[] tr : trust) {
                counter[tr[0]]--;
                counter[tr[1]]++;
            }
            for (int i = 1; i <= n; i++) {
                if (counter[i] == n - 1) return i;
            }
            return -1;
        }
    }

    static class SolutionBruteForce {
        public int findJudge(int n, int[][] trust) {
            Map<Integer, Set<Integer>> map = new HashMap<>();

            for (int i = 1; i <= n; i++) {
                map.put(i, new HashSet<>());
            }
            for (int[] pair : trust) {
                map.get(pair[0]).add(pair[1]);
            }
            int judge = possibleTownJudge(map);
            if (judge == -1) return judge;
            for (int i = 1; i <= n; i++) {
                if (!map.get(i).contains(judge) && i != judge) return -1;
            }
            return judge;
        }

        private int possibleTownJudge(Map<Integer, Set<Integer>> map) {
            for (Map.Entry<Integer, Set<Integer>> entry : map.entrySet()) {
                if (entry.getValue().isEmpty()) {
                    return entry.getKey();
                }
            }
            return -1;
        }
    }
}
